maintainers/scripts/haskell/hydra-report.hs

```diff
@@ -0,0 +1,233 @@
+import Data.Map (Map)
+import qualified Data.Map as Map
+import Data.Maybe (fromMaybe)
+import Data.Sequence (Seq)
+import qualified Data.Sequence as Seq
+import Data.Text (Text)
+import qualified Data.Text as Text
+import Data.Time (defaultTimeLocale, formatTime, getCurrentTime)
+import Data.Time.Clock (UTCTime)
+import GHC.Generics
+import Network.HTTP.Req
+import System.Environment (getArgs)
+import System.Process (readProcess)
+
+newtype JobsetEvals = JobsetEvals
+   { evals :: Seq Eval
+   }
+   deriving (Generic, ToJSON, FromJSON, Show)
+
+newtype Nixpkgs = Nixpkgs {revision :: Text}
+   deriving (Generic, ToJSON, FromJSON, Show)
+
+newtype JobsetEvalInputs = JobsetEvalInputs {nixpkgs :: Nixpkgs}
+   deriving (Generic, ToJSON, FromJSON, Show)
+
+data Eval = Eval
+   { builds :: Seq Int
+   , id :: Int
+   , jobsetevalinputs :: JobsetEvalInputs
+   }
+   deriving (Generic, ToJSON, FromJSON, Show)
+
+data Build = Build
+   { job :: Text
+   , buildstatus :: Int
+   , finished :: Int
+   , id :: Int
+   , nixname :: Text
+   , system :: Text
+   , jobsetevals :: Seq Int
+   }
+   deriving (Generic, ToJSON, FromJSON, Show)
+
+data BuildName = BuildName Text Int deriving (Show, Eq, Ord)
+
+data StatusSummary = StatusSummary
+   { successfulBuilds :: Int
+   , unfinishedBuilds :: Int
+   , canceledBuilds :: Int
+   , failedMaintainedBuilds :: Seq (BuildName, Seq Text)
+   , failedMaintainedDependencies :: Seq (BuildName, Seq Text)
+   , failedPackageBuilds :: Seq BuildName
+   , failedPackageDependencies :: Seq BuildName
+   , failedJobBuilds :: Seq BuildName
+   , failedJobDependencies :: Seq BuildName
+   }
+   deriving (Show)
+
+instance Semigroup StatusSummary where
+   (<>) lhs rhs =
+      StatusSummary
+         { successfulBuilds = successfulBuilds lhs + successfulBuilds rhs
+         , failedPackageBuilds = failedPackageBuilds lhs <> failedPackageBuilds rhs
+         , failedPackageDependencies = failedPackageDependencies lhs <> failedPackageDependencies rhs
+         , failedJobBuilds = failedJobBuilds lhs <> failedJobBuilds rhs
+         , failedJobDependencies = failedJobDependencies lhs <> failedJobDependencies rhs
+         , failedMaintainedBuilds = failedMaintainedBuilds lhs <> failedMaintainedBuilds rhs
+         , failedMaintainedDependencies = failedMaintainedDependencies lhs <> failedMaintainedDependencies rhs
+         , unfinishedBuilds = unfinishedBuilds lhs + unfinishedBuilds rhs
+         , canceledBuilds = canceledBuilds lhs + canceledBuilds rhs
+         }
+
+instance Monoid StatusSummary where
+   mempty = StatusSummary 0 0 0 mempty mempty mempty mempty mempty mempty
+
+main :: IO ()
+main = do
+   args <- getArgs
+   case args of
+      ["get-report"] -> getBuildReports
+      ["ping-maintainers"] -> printMaintainerPing
+      ["mark-broken-list"] -> printMarkBrokenList
+      _ -> putStrLn "Usage: get-report | ping-maintainers | mark-broken-list"
+
+reportFileName = "haskell-updates-build-report.json"
```

Comment:
Maybe we should save this in `XDG_CACHE_HOME` (fallback to `$HOME/.cache`).

Reply:
Good idea. Allthough that increases the dependency footprint of the script, I will do it.

Reply:
I don't think pulling in a xdg library is necessary really, you can just have two calls to `getEnv`.

Reply:
directory exports xdgDirectory and is bundled with ghc. So everything great here!